File: packages/python/handoffkit/evaluation.py
"""Offline workflow evaluation reports for HandoffKit."""

from __future__ import annotations

import json
from dataclasses import dataclass, field
from typing import Any

from handoffkit.handoff import HandoffState
from handoffkit.quality import HandoffQualityEvaluator
from handoffkit.recipes import RecipeRunResult
from handoffkit.replay import ReplayRunner
from handoffkit.schemas import TeamRunResult
from handoffkit.tool_execution import ToolExecutionReport, ToolResult
from handoffkit.tracing import RunTrace
from handoffkit.validation import HandoffStateValidator, ValidationReport


def _json_default(value: Any) -> Any:
    """Return readable JSON fallback for non-serializable values."""
    if hasattr(value, "to_dict"):
        return value.to_dict()
    return str(value)


@dataclass
class EvaluationCheck:
    """One named offline evaluation check."""

    name: str
    passed: bool
    message: str
    severity: str = "error"
    metadata: dict[str, Any] = field(default_factory=dict)

    def to_dict(self) -> dict[str, Any]:
        """Serialize this check."""
        return {
            "name": self.name,
            "passed": self.passed,
            "message": self.message,
            "severity": self.severity,
            "metadata": self.metadata,
        }


@dataclass
class EvaluationResult:
    """Result for one evaluated workflow artifact."""

    target: str
    success: bool
    checks: list[EvaluationCheck] = field(default_factory=list)
    score: float = 0.0
    metadata: dict[str, Any] = field(default_factory=dict)

    def to_dict(self) -> dict[str, Any]:
        """Serialize this result."""
        return {
            "target": self.target,
            "success": self.success,
            "checks": [check.to_dict() for check in self.checks],
            "score": round(self.score, 3),
            "metadata": self.metadata,
        }


@dataclass
class WorkflowEvaluationReport:
    """Reusable evaluation report for workflow readiness."""

    success: bool
    score: float
    grade: str
    results: list[EvaluationResult] = field(default_factory=list)
    recommendations: list[str] = field(default_factory=list)
    metadata: dict[str, Any] = field(default_factory=dict)

    def to_dict(self) -> dict[str, Any]:
        """Serialize this report."""
        return {
            "success": self.success,
            "score": round(self.score, 3),
            "grade": self.grade,
            "results": [result.to_dict() for result in self.results],
            "recommendations": self.recommendations,
            "metadata": self.metadata,
        }

    def to_json(self, *, indent: int | None = 2) -> str:
        """Serialize this report as JSON."""
        return json.dumps(self.to_dict(), ensure_ascii=False, indent=indent, default=_json_default)

    def to_markdown(self) -> str:
        """Serialize this report as Markdown."""
        results = "\n".join(
            (
                f"- `{result.target}` success={result.success} "
                f"score={result.score:.2f} checks={len(result.checks)}"
            )
            for result in self.results
        ) or "- none"
        checks = "\n".join(
            (
                f"- `{check.name}` passed={check.passed} "
                f"severity={check.severity}: {check.message}"
            )
            for result in self.results
            for check in result.checks
        ) or "- none"
        recommendations = "\n".join(f"- {item}" for item in self.recommendations) or "- none"
        return (
            "# Workflow Evaluation Report\n\n"
            f"## Success\n\n{self.success}\n\n"
            f"## Score\n\n{self.score:.3f} ({self.grade})\n\n"
            f"## Results\n\n{results}\n\n"
            f"## Checks\n\n{checks}\n\n"
            f"## Recommendations\n\n{recommendations}\n"
        )


class WorkflowEvaluator:
    """Evaluate workflow artifacts with deterministic offline checks."""

    def __init__(
        self,
        *,
        min_score: float = 0.75,
        handoff_min_score: float = 0.6,
    ) -> None:
        self.min_score = min_score
        self.handoff_min_score = handoff_min_score
        self.handoff_validator = HandoffStateValidator()
        self.quality_evaluator = HandoffQualityEvaluator(min_score=handoff_min_score)

    def evaluate(self, target: Any) -> WorkflowEvaluationReport:
        """Evaluate one workflow artifact."""
        if isinstance(target, RunTrace):
            result = self._evaluate_trace(target)
        elif isinstance(target, TeamRunResult):
            result = self._evaluate_team_result(target)
        elif isinstance(target, RecipeRunResult):
            result = self._evaluate_recipe_result(target)
        elif isinstance(target, ToolExecutionReport):
            result = self._evaluate_tool_report(target)
        elif self._is_handoff_list(target):
            result = self._evaluate_handoffs(list(target))
        else:
            result = EvaluationResult(
                target=target.__class__.__name__,
                success=False,
                checks=[
                    EvaluationCheck(
                        "supported_target",
                        False,
                        "target must be TeamRunResult, RecipeRunResult, RunTrace, "
                        "ToolExecutionReport, or a list of HandoffState",
                    )
                ],
                score=0.0,
            )
        return self._report([result])

    def evaluate_many(self, targets: list[Any]) -> WorkflowEvaluationReport:
        """Evaluate multiple workflow artifacts and aggregate results."""
        if not targets:
            return WorkflowEvaluationReport(
                success=False,
                score=0.0,
                grade="F",
                results=[],
                recommendations=["Provide at least one workflow artifact to evaluate."],
                metadata={"targets": 0, "evaluator": self.__class__.__name__},
            )
        results = [self.evaluate(target).results[0] for target in targets]
        return self._report(results)

    def from_report_json(self, payload: dict[str, Any]) -> WorkflowEvaluationReport:
        """Evaluate a serialized trace/report JSON object."""
        if {"run_id", "steps", "final_output"}.issubset(payload):
            return self.evaluate(RunTrace.from_dict(payload))
        if "handoffs" in payload and isinstance(payload.get("handoffs"), list):
            handoffs = [
                HandoffState.from_dict(item)
                for item in payload["handoffs"]
                if isinstance(item, dict)
            ]
            return self.evaluate(handoffs)
        return self.evaluate(payload)

    def _evaluate_team_result(self, result: TeamRunResult) -> EvaluationResult:
        trace = RunTrace.from_team_result(result, name="team-evaluation")
        checks = [
            EvaluationCheck(
                "team_output",
                bool(result.final_output),
                (
                    "team produced final output"
                    if result.final_output
                    else "team final output is empty"
                ),
            ),
            EvaluationCheck(
                "agent_outputs",
                bool(result.agent_outputs),
                "team recorded agent outputs"
                if result.agent_outputs
                else "team recorded no agent outputs",
            ),
        ]
        checks.extend(self._handoff_checks(result.handoffs))
        checks.extend(self._trace_checks(trace))
        return self._result("TeamRunResult", checks, {"handoffs": len(result.handoffs)})

    def _evaluate_recipe_result(self, result: RecipeRunResult) -> EvaluationResult:
        trace = RunTrace.from_recipe_result(result)
        checks = [
            EvaluationCheck(
                "recipe_success",
                result.success,
                "recipe run succeeded" if result.success else "recipe run failed",
            ),
            EvaluationCheck(
                "recipe_steps",
                bool(result.step_results),
                "recipe recorded step results"
                if result.step_results
                else "recipe recorded no step results",
            ),
        ]
        checks.extend(self._handoff_checks(result.handoff_states))
        checks.extend(self._tool_result_checks(result.tool_results))
        checks.extend(self._trace_checks(trace))
        return self._result("RecipeRunResult", checks, {"recipe": result.recipe_name})

    def _evaluate_trace(self, trace: RunTrace) -> EvaluationResult:
        checks = self._trace_checks(trace)
        checks.extend(self._handoff_checks(trace.handoffs))
        for step in trace.steps:
            checks.extend(self._tool_result_checks(step.tool_results))
        replay = ReplayRunner(trace).summary()
        checks.append(
            EvaluationCheck(
                "replay_summary",
                replay.step_count == len(trace.steps),
                "replay summary inspected trace without execution",
                metadata=replay.to_dict(),
            )
        )
        return self._result("RunTrace", checks, {"run_id": trace.run_id, "name": trace.name})

    def _evaluate_tool_report(self, report: ToolExecutionReport) -> EvaluationResult:
        checks = [
            EvaluationCheck(
                "tool_report_success",
                report.success,
                "tool report succeeded" if report.success else "tool report failed",
            ),
            EvaluationCheck(
                "tool_steps",
                bool(report.steps),
                "tool report recorded steps" if report.steps else "tool report recorded no steps",
            ),
        ]
        checks.extend(self._tool_result_checks(report.tool_results))
        return self._result("ToolExecutionReport", checks, {"agent": report.agent_name})

    def _evaluate_handoffs(self, handoffs: list[HandoffState]) -> EvaluationResult:
        checks = self._handoff_checks(handoffs)
        return self._result("HandoffStateList", checks, {"handoffs": len(handoffs)})

    def _handoff_checks(self, handoffs: list[Any]) -> list[EvaluationCheck]:
        if not handoffs:
            return [
                EvaluationCheck(
                    "handoffs_present",
                    False,
                    "no handoff states were recorded",
                    severity="warning",
                )
            ]
        checks = [
            EvaluationCheck(
                "handoffs_present",
                True,
                f"{len(handoffs)} handoff state(s) recorded",
                severity="info",
            )
        ]
        states = [item for item in handoffs if isinstance(item, HandoffState)]
        for index, state in enumerate(states):
            report = self.handoff_validator.validate(state)
            checks.append(self._validation_check(f"handoff_{index + 1}_contract", report))
        if states:
            quality = self.quality_evaluator.evaluate_many(states)
            checks.append(
                EvaluationCheck(
                    "handoff_quality",
                    quality.success,
                    f"handoff quality score {quality.score:.3f}",
                    metadata=quality.to_dict(),
                )
            )
        return checks

    def _trace_checks(self, trace: RunTrace) -> list[EvaluationCheck]:
        return [
            EvaluationCheck(
                "trace_success",
                trace.success,
                "trace succeeded" if trace.success else "trace indicates failure",
            ),
            EvaluationCheck(
                "trace_steps",
                bool(trace.steps),
                "trace recorded steps" if trace.steps else "trace recorded no steps",
            ),
            EvaluationCheck(
                "trace_final_output",
                bool(trace.final_output),
                "trace has final output" if trace.final_output else "trace final output is empty",
            ),
        ]

    def _tool_result_checks(self, results: list[Any]) -> list[EvaluationCheck]:
        checks = []
        for index, result in enumerate(results):
            if isinstance(result, ToolResult):
                checks.append(
                    EvaluationCheck(
                        f"tool_result_{index + 1}",
                        result.success,
                        (
                            f"tool {result.tool_name} succeeded"
                            if result.success
                            else f"tool {result.tool_name} failed: {result.error or 'unknown'}"
                        ),
                        metadata=result.to_dict(),
                    )
                )
        return checks

    def _validation_check(self, name: str, report: ValidationReport) -> EvaluationCheck:
        issues = "; ".join(issue.message for issue in report.errors)
        return EvaluationCheck(
            name,
            report.success,
            "contract validation passed" if report.success else issues or "contract failed",
            metadata=report.to_dict(),
        )

    def _result(
        self,
        target: str,
        checks: list[EvaluationCheck],
        metadata: dict[str, Any] | None = None,
    ) -> EvaluationResult:
        blocking = [check for check in checks if check.severity == "error"]
        passed = sum(1 for check in blocking if check.passed)
        score = passed / len(blocking) if blocking else 1.0
        success = score >= self.min_score and all(check.passed for check in blocking)
        return EvaluationResult(
            target=target,
            success=success,
            checks=checks,
            score=score,
            metadata=metadata or {},
        )

    def _report(self, results: list[EvaluationResult]) -> WorkflowEvaluationReport:
        score = sum(result.score for result in results) / len(results) if results else 0.0
        success = (
            bool(results)
            and all(result.success for result in results)
            and score >= self.min_score
        )
        recommendations = self._recommendations(results)
        return WorkflowEvaluationReport(
            success=success,
            score=score,
            grade=self._grade(score),
            results=results,
            recommendations=recommendations,
            metadata={
                "targets": len(results),
                "evaluator": self.__class__.__name__,
                "min_score": self.min_score,
                "handoff_min_score": self.handoff_min_score,
            },
        )

    def _recommendations(self, results: list[EvaluationResult]) -> list[str]:
        recommendations: list[str] = []
        for result in results:
            for check in result.checks:
                if check.passed or check.severity != "error":
                    continue
                if "handoff" in check.name:
                    recommendations.append("Improve handoff contracts and quality before release.")
                elif "trace" in check.name:
                    recommendations.append(
                        "Record complete run traces with steps and final output."
                    )
                elif "tool" in check.name:
                    recommendations.append("Resolve failed tool calls or tool execution reports.")
                elif "recipe" in check.name:
                    recommendations.append(
                        "Fix failing recipe steps before accepting the workflow."
                    )
                else:
                    recommendations.append(check.message)
        return list(dict.fromkeys(recommendations)) or [
            "Workflow artifacts passed offline evaluation."
        ]

    def _grade(self, score: float) -> str:
        if score >= 0.9:
            return "A"
        if score >= 0.75:
            return "B"
        if score >= 0.6:
            return "C"
        if score >= 0.4:
            return "D"
        return "F"

    def _is_handoff_list(self, target: Any) -> bool:
        return isinstance(target, list) and all(isinstance(item, HandoffState) for item in target)

File: packages/js/core/src/evaluation.js
import { agentName, grade, round, toJSONString } from "./utils.js";
import { HandoffState } from "./contracts.js";
import { TeamRunResult } from "./team.js";
import { HandoffQualityEvaluator } from "./quality.js";
import { ReplayRunner, RunTrace } from "./tracing.js";

export class EvaluationCheck {
  constructor({ name, passed, message, severity = "error", metadata = {} }) {
    this.name = name;
    this.passed = Boolean(passed);
    this.message = message;
    this.severity = severity;
    this.metadata = { ...metadata };
  }

  toJSON() {
    return {
      name: this.name,
      passed: this.passed,
      message: this.message,
      severity: this.severity,
      metadata: { ...this.metadata },
    };
  }
}

export class EvaluationResult {
  constructor({ target, success, checks = [], score = 0.0, metadata = {} }) {
    this.target = target;
    this.success = Boolean(success);
    this.checks = (Array.isArray(checks) ? checks : []).map(c => c instanceof EvaluationCheck ? c : new EvaluationCheck(c));
    this.score = Number(score);
    this.metadata = { ...metadata };
  }

  toJSON() {
    return {
      target: this.target,
      success: this.success,
      checks: this.checks.map(c => c.toJSON()),
      score: Math.round(this.score * 1000) / 1000,
      metadata: { ...this.metadata },
    };
  }
}

export class WorkflowEvaluationReport {
  constructor({ success, score, grade, results = [], recommendations = [], metadata = {} }) {
    this.success = Boolean(success);
    this.score = Number(score);
    this.grade = grade;
    this.results = (Array.isArray(results) ? results : []).map(r => r instanceof EvaluationResult ? r : new EvaluationResult(r));
    this.recommendations = Array.isArray(recommendations) ? [...recommendations] : [];
    this.metadata = { ...metadata };
  }

  toJSON() {
    return {
      success: this.success,
      score: Math.round(this.score * 1000) / 1000,
      grade: this.grade,
      results: this.results.map(r => r.toJSON()),
      recommendations: [...this.recommendations],
      metadata: { ...this.metadata },
    };
  }

  toJSONString(space = 2) {
    return JSON.stringify(this.toJSON(), null, space);
  }

  toMarkdown() {
    const results = this.results.map(
      r => `- \`${r.target}\` success=${r.success} score=${r.score.toFixed(2)} checks=${r.checks.length}`
    ).join("\n") || "- none";

    const checks = this.results.flatMap(r => r.checks).map(
      c => `- \`${c.name}\` passed=${c.passed} severity=${c.severity}: ${c.message}`
    ).join("\n") || "- none";

    const recs = this.recommendations.map(r => `- ${r}`).join("\n") || "- none";

    return [
      "# Workflow Evaluation Report",
      "",
      "## Success",
      "",
      String(this.success),
      "",
      "## Score",
      "",
      `${this.score.toFixed(3)} (${this.grade})`,
      "",
      "## Results",
      "",
      results,
      "",
      "## Checks",
      "",
      checks,
      "",
      "## Recommendations",
      "",
      recs
    ].join("\n");
  }
}

export class WorkflowEvaluator {
  constructor({ minScore = 0.75, handoffMinScore = 0.6 } = {}) {
    this.minScore = minScore;
    this.handoffMinScore = handoffMinScore;
  }

  evaluate(target) {
    let result;
    if (target instanceof RunTrace) {
      result = this._evaluateTrace(target);
    } else if (target instanceof TeamRunResult) {
      result = this._evaluateTeamResult(target);
    } else if (target && (target.recipeName || target.recipe_name)) {
      result = this._evaluateRecipeResult(target);
    } else if (target && (target.toolResults || target.tool_results) && target.agentName) {
      result = this._evaluateToolReport(target);
    } else if (Array.isArray(target) && target.every(item => item instanceof HandoffState)) {
      result = this._evaluateHandoffs(target);
    } else {
      result = new EvaluationResult({
        target: target?.constructor?.name || "unknown",
        success: false,
        checks: [
          new EvaluationCheck({
            name: "supported_target",
            passed: false,
            message: "target must be TeamRunResult, RecipeRunResult, RunTrace, ToolExecutionReport, or an array of HandoffState",
          })
        ],
        score: 0.0,
      });
    }
    return this._report([result]);
  }

  evaluateMany(targets) {
    if (!Array.isArray(targets) || targets.length === 0) {
      return new WorkflowEvaluationReport({
        success: false,
        score: 0.0,
        grade: "F",
        results: [],
        recommendations: ["Provide at least one workflow artifact to evaluate."],
        metadata: { targets: 0, evaluator: this.constructor.name },
      });
    }
    const results = targets.map(t => this.evaluate(t).results[0]);
    return this._report(results);
  }

  _evaluateTeamResult(result) {
    const trace = RunTrace.fromTeamResult(result, { name: "team-evaluation" });
    const checks = [
      new EvaluationCheck({
        name: "team_output",
        passed: Boolean(result.finalOutput),
        message: result.finalOutput ? "team produced final output" : "team final output is empty",
      }),
      new EvaluationCheck({
        name: "agent_outputs",
        passed: Array.isArray(result.stepResults) && result.stepResults.length > 0,
        message: Array.isArray(result.stepResults) && result.stepResults.length > 0 ? "team recorded agent outputs" : "team recorded no agent outputs",
      }),
    ];
    checks.push(...this._handoffChecks(result.handoffs));
    checks.push(...this._traceChecks(trace));
    return this._result("TeamRunResult", checks, { handoffs: result.handoffs?.length ?? 0 });
  }

  _evaluateRecipeResult(result) {
    const trace = RunTrace.fromTeamResult(result, { name: "recipe-evaluation" });
    const checks = [
      new EvaluationCheck({
        name: "recipe_success",
        passed: result.success,
        message: result.success ? "recipe run succeeded" : "recipe run failed",
      }),
      new EvaluationCheck({
        name: "recipe_steps",
        passed: Array.isArray(result.stepResults) && result.stepResults.length > 0,
        message: Array.isArray(result.stepResults) && result.stepResults.length > 0 ? "recipe recorded step results" : "recipe recorded no step results",
      }),
    ];
    const handoffs = result.handoffs || result.handoffStates || [];
    checks.push(...this._handoffChecks(handoffs));
    checks.push(...this._traceChecks(trace));
    return this._result("RecipeRunResult", checks, { recipe: result.recipeName || result.recipe_name });
  }

  _evaluateTrace(trace) {
    const checks = this._traceChecks(trace);
    checks.push(...this._handoffChecks(trace.handoffs));
    for (const step of trace.steps) {
      if (step.toolResults) {
        checks.push(...this._toolResultChecks(step.toolResults));
      }
    }
    const replay = new ReplayRunner(trace).summary();
    checks.push(
      new EvaluationCheck({
        name: "replay_summary",
        passed: replay.stepCount === trace.steps.length,
        message: "replay summary inspected trace without execution",
        metadata: replay.toJSON ? replay.toJSON() : replay,
      })
    );
    return this._result("RunTrace", checks, { runId: trace.runId, name: trace.name });
  }

  _evaluateToolReport(report) {
    const checks = [
      new EvaluationCheck({
        name: "tool_report_success",
        passed: report.success,
        message: report.success ? "tool report succeeded" : "tool report failed",
      }),
      new EvaluationCheck({
        name: "tool_steps",
        passed: Array.isArray(report.steps) && report.steps.length > 0,
        message: Array.isArray(report.steps) && report.steps.length > 0 ? "tool report recorded steps" : "tool report recorded no steps",
      }),
    ];
    const toolResults = report.toolResults || report.tool_results || [];
    checks.push(...this._toolResultChecks(toolResults));
    return this._result("ToolExecutionReport", checks, { agent: report.agentName || report.agent_name });
  }

  _evaluateHandoffs(handoffs) {
    const checks = this._handoffChecks(handoffs);
    return this._result("HandoffStateList", checks, { handoffs: handoffs.length });
  }

  _handoffChecks(handoffs) {
    if (!Array.isArray(handoffs) || handoffs.length === 0) {
      return [
        new EvaluationCheck({
          name: "handoffs_present",
          passed: false,
          message: "no handoff states were recorded",
          severity: "warning",
        })
      ];
    }
    const checks = [
      new EvaluationCheck({
        name: "handoffs_present",
        passed: true,
        message: `${handoffs.length} handoff state(s) recorded`,
        severity: "info",
      })
    ];
    for (let index = 0; index < handoffs.length; index++) {
      const state = handoffs[index];
      const validation = state.validate ? state.validate() : state;
      const passed = !validation.errors || validation.errors.length === 0;
      checks.push(
        new EvaluationCheck({
          name: `handoff_${index + 1}_contract`,
          passed,
          message: passed ? "contract validation passed" : (validation.errors?.join("; ") || "contract failed"),
          metadata: state.toJSON ? state.toJSON() : state,
        })
      );
    }
    const evaluator = new HandoffQualityEvaluator({ minScore: this.handoffMinScore });
    const quality = evaluator.evaluateMany ? evaluator.evaluateMany(handoffs) : { success: true, score: 1.0 };
    checks.push(
      new EvaluationCheck({
        name: "handoff_quality",
        passed: quality.success,
        message: `handoff quality score ${(quality.score ?? 1.0).toFixed(3)}`,
        metadata: quality.toJSON ? quality.toJSON() : quality,
      })
    );
    return checks;
  }

  _traceChecks(trace) {
    return [
      new EvaluationCheck({
        name: "trace_success",
        passed: trace.success,
        message: trace.success ? "trace succeeded" : "trace indicates failure",
      }),
      new EvaluationCheck({
        name: "trace_steps",
        passed: Array.isArray(trace.steps) && trace.steps.length > 0,
        message: Array.isArray(trace.steps) && trace.steps.length > 0 ? "trace recorded steps" : "trace recorded no steps",
      }),
      new EvaluationCheck({
        name: "trace_final_output",
        passed: Boolean(trace.finalOutput),
        message: trace.finalOutput ? "trace has final output" : "trace final output is empty",
      }),
    ];
  }

  _toolResultChecks(results) {
    const checks = [];
    const list = Array.isArray(results) ? results : [];
    for (let index = 0; index < list.length; index++) {
      const result = list[index];
      const toolName = result.name || result.tool_name || "unknown";
      checks.push(
        new EvaluationCheck({
          name: `tool_result_${index + 1}`,
          passed: result.success !== false,
          message: result.success !== false ? `tool ${toolName} succeeded` : `tool ${toolName} failed: ${result.error || "unknown"}`,
          metadata: result.toJSON ? result.toJSON() : result,
        })
      );
    }
    return checks;
  }

  _result(target, checks, metadata = {}) {
    const blocking = checks.filter(c => c.severity === "error");
    const passed = blocking.filter(c => c.passed).length;
    const score = blocking.length > 0 ? passed / blocking.length : 1.0;
    const success = score >= this.minScore && blocking.every(c => c.passed);
    return new EvaluationResult({ target, success, checks, score, metadata });
  }

  _report(results) {
    const score = results.length > 0 ? results.reduce((sum, r) => sum + r.score, 0) / results.length : 0.0;
    const success = results.length > 0 && results.every(r => r.success) && score >= this.minScore;
    const recommendations = this._recommendations(results);
    return new WorkflowEvaluationReport({
      success,
      score,
      grade: this._grade(score),
      results,
      recommendations,
      metadata: {
        targets: results.length,
        evaluator: this.constructor.name,
        minScore: this.minScore,
        handoffMinScore: this.handoffMinScore,
      },
    });
  }

  _recommendations(results) {
    const recs = [];
    for (const result of results) {
      for (const check of result.checks) {
        if (check.passed || check.severity !== "error") continue;
        if (check.name.includes("handoff")) {
          recs.push("Improve handoff contracts and quality before release.");
        } else if (check.name.includes("trace")) {
          recs.push("Record complete run traces with steps and final output.");
        } else if (check.name.includes("tool")) {
          recs.push("Resolve failed tool calls or tool execution reports.");
        } else if (check.name.includes("recipe")) {
          recs.push("Fix failing recipe steps before accepting the workflow.");
        } else {
          recs.push(check.message);
        }
      }
    }
    const unique = [...new Set(recs)];
    return unique.length > 0 ? unique : ["Workflow artifacts passed offline evaluation."];
  }

  _grade(score) {
    if (score >= 0.9) return "A";
    if (score >= 0.75) return "B";
    if (score >= 0.6) return "C";
    if (score >= 0.4) return "D";
    return "F";
  }
}

